Throw an exception (mark as failed) scheduled actions without any callbacks

barryhughes opened the pull request:
This change is geared toward handling of scheduled actions that do not have any callbacks and, more broadly, recurring actions that continually and consistently fail.

Example: imagine a plugin has scheduled an hourly recurring action, but that same plugin is subsequently deleted without any opportunity to perform clean-up (or, it is deactivated normally but does not contain uninstall logic). Currently, those recurring actions will continue to fire and:

- Each time the action status will be updated to `complete` (so it will seem like they are running successfully).
- A replacement action will be scheduled.

Therefore, we can end up with a category of repeating zombie actions, clogging up the queue unnecessarily. This change attempts to address that problem and closes #805.

### How it works

Just because an action fails once doesn't mean it will always fail: so, we need to bake in some amount of tolerance. The way I propose we handle this is two-fold:

- When an action does not have any callbacks, mark it as `failed`.
- If it is a recurring action, check to see if it is *consistently* failing and, if so, don't reschedule it.
    - By default, the criteria for *"consistent failure"* is when the most recent 5 actions with the same hook, argument set, and schedule have all failed.
    - This will be adjustable via a filter.

### Testing

To make testing easier, I suggest temporarily stopping Action Scheduler from being triggered automatically:

- Add `define( 'DISABLE_WP_CRON', true );` to your `wp-config.php` file.
- Add `add_filter( 'action_scheduler_allow_async_request_runner', '__return_false' );` to a suitable location, such as a mu-plugin file.
- If you have any crontab entries that provide alternative ways of running cron or the Action Scheduler queue, suspend them.
- Also, to give yourself a 'clean view' of things, you may also wish to delete all existing actions: `wp db query "DELETE FROM $(wp db prefix)actionscheduler_actions"`

Now, processing of the queue should be entirely manual. Next:

- Create a new recurring action without a callback: `wp eval "as_schedule_recurring_action( time() - 1, 2, 'no_callback' );"`
- Visit **Tools ▸ Scheduled Actions** and locate action `no_callback` (should be pending), then run it manually.
- It should fail, and a replacement should be scheduled.
- Rinse and repeat a few times.
- In the end, you should find you have 5 failed scheduled actions for the hook `no_callback` and no more pending actions of that name should have been scheduled.

### Notes

- This is one facet of an effort to address zombie actions; we need to follow-up with further work that prevents unintentional duplication of actions (via race conditions).
- You will notice a large number of updates to tests that seem to be unrelated: that's because those tests relied on the previous behavior of Action Scheduler, where actions would be marked complete even if they didn't have a callback. Most of these updates are isolated inside commits https://github.com/woocommerce/action-scheduler/pull/806/commits/8276036dd57d3ba8ec8249e55a386c5acb1bfc4c and https://github.com/woocommerce/action-scheduler/pull/806/commits/37b8a1efb8911efc657609c0aa11c3392e3a4b86. The assertions specific to the core change can be [found here](https://github.com/woocommerce/action-scheduler/pull/806/files#diff-2fd59945ebe22bbea56dd19195fec979d5ee027dc705ac6b6af4eb01defad288).

### Changelog

> Enhancement - If a recurring action is found to be consistently failing, it will stop being rescheduled.

barryhughes committed:
Throw an exception (mark as failed) scheduled actions without any callbacks.

barryhughes committed:
Correct further tests referencing hooks-without-callbacks.

barryhughes committed:
Refactor stub callbacks for tests (support deprecated/legacy test envs).

barryhughes committed:
Safeguards to avoid rescheduling a recurring action that is persistently failing (WIP).

barryhughes commented:
Started adding code to guard against rescheduling a consistently failing recurring action. Need to add additional test-cases.

barryhughes committed:
Improve logic for detecting consistently-failing-recurring actions.

vedanshujain commented on classes/abstracts/ActionScheduler_Abstract_QueueRunner.php:
```diff
@@ -86,13 +86,94 @@ public function process_action( $action_id, $context = '' ) {
+	 * @return bool
+	 */
+	private function recurring_action_is_consistently_failing( ActionScheduler_Action $action, $action_id ) {
+		// Fetch the most recent historic actions having the same hook and argument set.
```
May I suggest a bit more performant algorithm to solve this method:
1. Get the last `n`th action ID that failed for the same hook and args, where n is `action_scheduler_recurring_action_failure_threshold` using `per_page = 1`, `status = failed` and `offset = n`.  If there isn't any then`return false`. Store this action ID somewhere.
2. Get the last `n`th action ID with the same hook and args, but this time don't pass the `status = failed` filter. If the last n actions have been consistently failing, then both these ID have to be the same. So return true if this ID is the same as the previous one otherwise return false.

This will be more performant than the current query because you can do the comparison in just 2 queries, instead of `n+1` queries like we do now (and hopefully bit less code). This won't take schedule into account, but I think that's fine as we use the method in deciding whether to schedule another recurring action or not (so we know that the current action is recurring for sure).

barryhughes replied:
@vedanshujain—that's a great idea, thank you.

> This won't take schedule into account, but I think that's fine as we use the method in deciding whether to schedule another recurring action or not (so we know that the current action is recurring for sure).

My only nit would be that, with this strategy, we don't know if all of those failing actions were recurring—we will only know that they have the same hook (ie, this algorithm does not account for a mix of single and recurring actions using the same hook).

Even so, it is probably still a very reasonable way to go. Talking it out some more:

- The specifics of the schedule (beyond the fact we're primarily interested in recurring actions) probably don't matter all that much: after all, when an action is triggered, the callback doesn't know anything about the schedule—it just knows what its arguments are.
- Taking this a step further, I am now wondering if checking that the args match is actually the best policy, because it means we will fail to detect and halt a failing sequence like the following:

| Hook | Args | Status |
| --- | --- | --- |
| foo | [ 'page' => 1 ] | failed |
| foo | [ 'page' => 2 ] | failed |
| foo | [ 'page' => 3 ] | failed |
| ... | ... | ... |

To that end, I think I will:

- Adopt the strategy you proposed
- Perform the matching based only on the hook and not the args

Let me know if you have thoughts on this!

barryhughes replied:
Updated in https://github.com/woocommerce/action-scheduler/pull/806/commits/75b2766085b614d79699acfaacd29a1617cc1943

vedanshujain reviewed (commented):
Great work, I have added an optional suggestion but looks pretty good.

barryhughes committed:
Optimize algorithm for determining consistent failure; document filter hook.

barryhughes committed:
Tidy docblock.

barryhughes committed:
Revise threshold calculation; update test to match new logic.

barryhughes committed:
Drop method used to determine if schedules match.

After consideration, this felt unnecessary (callbacks have no knowledge as to schedule, so it seems more important to detect that a specific hook results in consistent failure (and limiting to the same args+schedule could therefore still allow failing actions to propagate) propagation.

barryhughes commented:
Thanks @vedanshujain. Ready for another look! Principal changes are [here (algorithm)](https://github.com/woocommerce/action-scheduler/pull/806/files#diff-63b59dbe3f432227cb577929975389b1b2b929d1a8adbbf2ac0e2517d7e10066R109-R152) and [here (assertions)](https://github.com/woocommerce/action-scheduler/pull/806/files#diff-2fd59945ebe22bbea56dd19195fec979d5ee027dc705ac6b6af4eb01defad288R192-R243):

- Hopefully a more efficient algorithm per your suggestion.
- No longer matches on arguments or on schedule: reasoning is that if a callback for `some_hook` consistently fails, we want to stop it from propagating. Precisely matching the args and schedule set the bar to a level that may make this preventative measure less effective.

vedanshujain reviewed (approved):
Looks neat!!
